wandb/sdk/launch/launch_add.py

```diff
@@ -91,14 +101,28 @@ def _launch_add(
         launch_config,
         cuda,
     )
-
+    if (
+        launch_spec.get("uri") is None
+        and launch_spec.get("job") is None
+        and launch_spec.get("docker", {}).get("docker_image") is None
+    ):
+        raise ValueError("Must specify either uri or job or docker_image")
```

Comment:
I feel like this gets done enough throughout the PR that you could add a utility called `validate_launch_spec` or something like that, which performs this check. You could then also create a unit test for it and save yourself all this code cov spam.

Reply:
yeah, I'll put it on LaunchProject init, and remove other references to them

Reply:
Oo actually it will need to be on the spec. okay!